Add browser target for binaryen toolchain utilities (#6131)
WebAssembly/binaryen: CMakeLists.txt

    add_compile_flag("-sDISABLE_EXCEPTION_CATCHING=0")
    add_link_flag("-sDISABLE_EXCEPTION_CATCHING=0")
  endif()
  # make the tools immediately usable on Node.js, unless building for the browser
  if (BUILD_FOR_BROWSER)
    add_link_flag("-sENVIRONMENT=web,worker")

>     add_link_flag("-sINVOKE_RUN=0")

Reviewer: Maybe add a comment about this flag? Something like "In the browser there is no natural place to provide commandline arguments for a commandline tool, so let the user run the main entry point themselves and pass in the arguments there."

    add_link_flag("-sMODULARIZE")
    add_link_flag("-sEXPORT_ES6")
    add_link_flag("-sFILESYSTEM")
    add_link_flag("-sFORCE_FILESYSTEM")

>     add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")

Reviewer: Can move this next to `INVOKE_RUN` as it is related (and the comment could cover both).

Author: Addressed in https://github.com/WebAssembly/binaryen/pull/6131/commits/5057ddd37d0a432f3203d2d5697bde10e16f64da.

>   # make the tools immediately usable on Node.js, unless building for the browser
>   if (BUILD_FOR_BROWSER)
>     add_link_flag("-sENVIRONMENT=web,worker")
>     add_link_flag("-sINVOKE_RUN=0")
>     add_link_flag("-sMODULARIZE")
>     add_link_flag("-sEXPORT_ES6")
>     add_link_flag("-sFILESYSTEM")
>     add_link_flag("-sFORCE_FILESYSTEM")
>     add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")
>   else()
>     add_link_flag("-sNODERAWFS")

Author: ```suggestion
  # In the browser, there is no natural place to provide commandline arguments
  # for a commandline tool, so let the user run the main entry point themselves
  # and pass in the arguments there.
  if (BUILD_FOR_BROWSER)
    add_link_flag("-sENVIRONMENT=web,worker")
    add_link_flag("-sINVOKE_RUN=0")
    add_link_flag("-sMODULARIZE")
    add_link_flag("-sEXPORT_ES6")
    add_link_flag("-sFILESYSTEM")
    add_link_flag("-sFORCE_FILESYSTEM")
    add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")
    # On Node.js, make the tools immediately usable.
    else()
    add_link_flag("-sNODERAWFS")
```

    add_compile_flag("-sDISABLE_EXCEPTION_CATCHING=0")
    add_link_flag("-sDISABLE_EXCEPTION_CATCHING=0")
  endif()
  # make the tools immediately usable on Node.js, unless building for the browser

>   if (NOT BUILD_FOR_BROWSER)
>     add_link_flag("-sNODERAWFS")
>   endif()
>   if(BUILD_FOR_BROWSER)
>     add_link_flag("-sENVIRONMENT=web,worker")
>     add_link_flag("-sINVOKE_RUN=0")
>     add_link_flag("-sMODULARIZE")
>     add_link_flag("-sEXPORT_ES6")
>     add_link_flag("-sFILESYSTEM")
>     add_link_flag("-sFORCE_FILESYSTEM")
>     add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")
>   endif()

Author: ```suggestion
  if (BUILD_FOR_BROWSER)
    add_link_flag("-sENVIRONMENT=web,worker")
    add_link_flag("-sINVOKE_RUN=0")
    add_link_flag("-sMODULARIZE")
    add_link_flag("-sEXPORT_ES6")
    add_link_flag("-sFILESYSTEM")
    add_link_flag("-sFORCE_FILESYSTEM")
    add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")
  else()
    add_link_flag("-sNODERAWFS")
  endif()
```

    add_compile_flag("-sDISABLE_EXCEPTION_CATCHING=0")
    add_link_flag("-sDISABLE_EXCEPTION_CATCHING=0")
  endif()
  # make the tools immediately usable on Node.js, unless building for the browser
  if (NOT BUILD_FOR_BROWSER)
    add_link_flag("-sNODERAWFS")
  endif()

>   if(BUILD_FOR_BROWSER)

Reviewer 2: nit: Can you write this as "if" + "else" instead of two ifs?

Author: Addressed in https://github.com/WebAssembly/binaryen/pull/6131/commits/018f0aa45b1bc6a29d07bb3fc4c3d651880b36e6.

    add_link_flag("-sENVIRONMENT=web,worker")
    add_link_flag("-sINVOKE_RUN=0")
    add_link_flag("-sNODERAWFS=0")
    add_link_flag("-sMODULARIZE=1")
    add_link_flag("-sEXPORT_ES6")
    add_link_flag("-sFILESYSTEM=1")
    add_link_flag("-sFORCE_FILESYSTEM")

>     add_link_flag("-sSINGLE_FILE=0")

Reviewer 2: nits:

You don't need `-sSINGLE_FILE=0` or `-sNODERAWFS=0` since they are the default.

Also, you can drop the `=1` from MODULARIZE and FILESYSTEM

Author: Addressed in https://github.com/WebAssembly/binaryen/pull/6131/commits/246ea0ec125f603cbbbe599b114ae0ffb02524b9.

>     add_link_flag("-sENVIRONMENT=web,worker")
>     add_link_flag("-sINVOKE_RUN=0")
>     add_link_flag("-sNODERAWFS=0")
>     add_link_flag("-sMODULARIZE=1")
>     add_link_flag("-sEXPORT_ES6")
>     add_link_flag("-sFILESYSTEM=1")
>     add_link_flag("-sFORCE_FILESYSTEM")
>     add_link_flag("-sSINGLE_FILE=0")

Author: ```suggestion
    add_link_flag("-sENVIRONMENT=web,worker")
    add_link_flag("-sINVOKE_RUN=0")
    add_link_flag("-sMODULARIZE")
    add_link_flag("-sEXPORT_ES6")
    add_link_flag("-sFILESYSTEM")
    add_link_flag("-sFORCE_FILESYSTEM")
```

>     add_link_flag("-sINVOKE_RUN=0")
>     add_link_flag("-sMODULARIZE")
>     add_link_flag("-sEXPORT_ES6")
>     add_link_flag("-sFILESYSTEM")
>     add_link_flag("-sFORCE_FILESYSTEM")
>     add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")

Author: ```suggestion
    add_link_flag("-sINVOKE_RUN=0")
    add_link_flag("-sEXPORTED_RUNTIME_METHODS=run,callMain,FS")    
    add_link_flag("-sMODULARIZE")
    add_link_flag("-sEXPORT_ES6")
    add_link_flag("-sFILESYSTEM")
    add_link_flag("-sFORCE_FILESYSTEM")
```

    add_link_flag("-sNODERAWFS")

>   endif()

Reviewer: Indentation looks wrong on these last three lines?

Author: Ooops, fixed in https://github.com/WebAssembly/binaryen/pull/6131/commits/049abfcbd1c77a6f24e8f99d94bf37c0b6d0c8db.
